File: ZigzagIterator.java

import java.util.List;
import java.util.ArrayList;

public class ZigzagIterator {

    List<List<Integer>> vectors;
    int[] indexes;
    int indexVectors;
    int endedVectors;

    public ZigzagIterator(List<Integer> v1, List<Integer> v2) {
        vectors = new ArrayList<>();
        vectors.add(v1);
        vectors.add(v2);
        indexes = new int[vectors.size()];
    }

    public int next() {
        if (!hasNext()) {
            throw new IndexOutOfBoundsException(indexes[indexVectors] + " of vector " + indexVectors + " is out of bounds!");
        }

        int next = vectors.get(indexVectors).get(indexes[indexVectors]);
        indexes[indexVectors]++;
        getNextIndexVectors();

        return next;
    }

    public boolean hasNext() {
        moveForward();
        return endedVectors < vectors.size();
    }

    public void moveForward() {
        endedVectors = 0;
        while (indexes[indexVectors] == vectors.get(indexVectors).size() && endedVectors < vectors.size()) {
            endedVectors++;
            getNextIndexVectors();
        }
    }

    public void getNextIndexVectors() {
        indexVectors = (indexVectors + 1) % vectors.size();
    }
}

File: ZigzagIterator.js

/**
 * @constructor
 * @param {Integer[]} v1
 * @param {Integer[]} v2
 */
var ZigzagIterator = function ZigzagIterator(v1, v2) {
    this.vectors = [v1, v2];
    this.indexes = new Array(this.vectors.length).fill(0);
    this.indexVectors = 0;
    this.endedVectors = 0;
};

/**
 * @this ZigzagIterator
 * @returns {integer}
 */
ZigzagIterator.prototype.next = function next() {
    if (!this.hasNext()) {
        throw `${indexes[indexVectors]} of vector ${indexVectors} is out of bounds!"`;
    }

    let next = this.vectors[this.indexVectors][this.indexes[this.indexVectors]];
    this.indexes[this.indexVectors]++;
    this.getNextIndexAmongLists();

    return next;
};


/**
 * @this ZigzagIterator
 * @returns {boolean}
 */
ZigzagIterator.prototype.hasNext = function hasNext() {
    this.moveForward();
    return this.endedVectors < this.vectors.length;
};

ZigzagIterator.prototype.moveForward = function () {
    this.endedVectors = 0;
    while (this.indexes[this.indexVectors] === this.vectors[this.indexVectors].length && this.endedVectors < this.vectors.length) {
        this.endedVectors++;
        this.getNextIndexAmongLists();
    }
};

ZigzagIterator.prototype.getNextIndexAmongLists = function () {
    this.indexVectors = (this.indexVectors + 1) % this.vectors.length;
};
